metronome/core/src/io/iohk/metronome/core/messages/RPCTracker.scala

```diff
@@ -0,0 +1,102 @@
+package io.iohk.metronome.core.messages
+
+import cats.implicits._
+import cats.effect.{Concurrent, Timer, Sync}
+import cats.effect.concurrent.{Ref, Deferred}
+import java.util.UUID
+import scala.concurrent.duration.FiniteDuration
+import scala.reflect.ClassTag
+
+/** `RPCTracker` can be used to register outgoing requests and later
+  * match them up with incoming responses, thus it facilitates turning
+  * the two independent messages into a `Kleisli[F, Request, Option[Response]]`,
+  * by a component that has access to the network, where a `None` result means
+  * the operation timed out before a response was received.
+  *
+  * The workflow is:
+  * 0. Receive some request parameters in a method.
+  * 1. Create a request ID.
+  * 2. Create a request with the ID.
+  * 3. Register the request with the tracker, hold on to the handle.
+  * 4. Send the request over the network.
+  * 5. Wait on the handle, eventually returning the optional result to the caller.
+  * 6. Pass every response received from the network to the tracker (on the network handler fiber).
+  */
+class RPCTracker[F[_]: Timer: Concurrent, M](
+    deferredMapRef: Ref[F, Map[UUID, RPCTracker.Entry[F, _]]],
+    defaultTimeout: FiniteDuration
+) {
+  import RPCTracker.Entry
+
+  def register[
+      Req <: RPCMessageCompanion#Request,
+      Res <: RPCMessageCompanion#Response
+  ](
+      request: Req,
+      timeout: FiniteDuration = defaultTimeout
+  )(implicit
+      ev1: Req <:< M,
+      ev2: RPCPair.Aux[Req, Res],
+      // Used by `RPCTracker.Entry.complete` to make sure only the
+      // expected response type can complete a request.
+      ct: ClassTag[Res]
+  ): F[F[Option[Res]]] = {
+    val requestId = request.requestId
+    for {
+      d <- Deferred[F, Option[Res]]
+      e = RPCTracker.Entry(d)
+      _ <- deferredMapRef.update(_ + (requestId -> e))
+      _ <- Concurrent[F].start {
+        Timer[F].sleep(timeout) >> completeWithTimeout(requestId)
+      }
+    } yield d.get
+  }
+
+  def complete[Res <: RPCMessageCompanion#Response](
+      response: Res
+  )(implicit ev: Res <:< M): F[Boolean] = {
+    remove(response.requestId).flatMap {
+      case None    => false.pure[F]
+      case Some(e) => e.complete(response)
+    }
+  }
+
+  private def completeWithTimeout(requestId: UUID): F[Unit] =
+    remove(requestId).flatMap {
+      case None    => ().pure[F]
+      case Some(e) => e.timeout
+    }
+
+  private def remove(requestId: UUID): F[Option[Entry[F, _]]] =
+    deferredMapRef.modify { dm =>
+      (dm - requestId, dm.get(requestId))
+    }
+}
+
+object RPCTracker {
+  case class Entry[F[_]: Sync, Res](
+      deferred: Deferred[F, Option[Res]]
+  )(implicit ct: ClassTag[Res]) {
+    def timeout: F[Unit] =
+      deferred.complete(None).attempt.void
```

Comment:
Perhaps a timeout is worth a tracer? Will we be able to add it?

Reply:
I don't want to add it to the `RPCTracker` itself, because in `core` there's no dependency on `tracing`, and currently tracers are on a bit coarser granularity (service level), but there's indeed [a trace event](https://github.com/input-output-hk/metronome/blob/1447f0bb0990d751223fe1e24d9c8691658a3fc4/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/SyncService.scala#L102) for timeouts in the `SyncService`.

Reply:
There's also one for [ignored responses](https://github.com/input-output-hk/metronome/blob/1447f0bb0990d751223fe1e24d9c8691658a3fc4/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/SyncService.scala#L159) so you can investigate why a certain response type always times out and always ignored, _and_ at the same time -> conclude that it's the wrong type.